### PR TITLE
Align DASC analysis and lifecycle contracts

### PR DESCRIPTION
**Pull request by kaix-nv**
Addresses the follow-up review findings on #2388.

Changes:
- expose optional `decay_parameter_storage_dtype` canonicalization through public `analyze_gdn_decay`, so callers can derive exactly the geometry consumed by calibration
- document that evaluated masks and measurement head counts must use the same declared storage dtype
- make unsupported-architecture save and restore both fail closed
- keep GDN geometry/decay drift serializable with a warning and deployment-blocked until recalibration
- add public analysis and save/restore symmetry regressions

Validation:
- focused DASC tests: 24 passed, 1 skipped (optional Megatron dependency)
- state + weight + attention sparsity tests: 302 passed, 1 skipped
- pre-commit on all touched files: passed

Commit is ED25519-signed and carries a matching Signed-off-by trailer.

## Summary by CodeRabbit

- **New Features**
  - Calibration now accounts for checkpoint storage formats (float16, bfloat16, and float32) when analyzing decay behavior.
  - Redundant precision-rounding slack is excluded when storage formats are compatible.
  - Invalid decay parameters and storage-format selections are clearly rejected.

- **Bug Fixes**
  - Recoverable structure or decay mismatches now allow metadata updates to continue with warnings.
  - Checkpoint serialization fails safely when a calibrated GDN layer is replaced with an unsupported module.

- **Documentation**
  - Clarified calibration parameter requirements and sequential handling of rounding bounds.

### CODE REVIEW

File: docs/source/guides/6_sparsity.rst [hunk 1]
```diff
@@ -146,6 +146,12 @@ ordinary dense recurrent state before continuation.
         "model_config_id": "sha256:<config-digest>",
         "calibration_data_id": "sha256:<dataset-and-protocol-digest>",
     }
+    # Use these storage-canonical horizons to derive the evaluated mask and the
+    # retained_heads/total_heads measurement geometry for every Wmax candidate.
+    horizons = mtss.analyze_gdn_decay(
+        model,
+        decay_parameter_storage_dtype=config["decay_parameter_storage_dtype"],
+    )
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply docs/source/guides/6_sparsity.rst:158]: Valid. Fixed in #2391. The documented config now makes epsilon and static_gate_input explicit and forwards them, along with decay_parameter_storage_dtype, to analyze_gdn_decay. The prose calls out all three as the measurement-geometry contract.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 2]
```diff
@@ -215,10 +216,23 @@ def analyze_gdn_decay(
     *,
     epsilon: float = 1e-3,
     static_gate_input: float = -0.3,
+    decay_parameter_storage_dtype: Literal["float16", "bfloat16", "float32"] | None = None,
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Addressed in #2391. A single internal _DecayParameterStorageDtype alias now types both config schemas, the public analysis argument, and the runtime mapping keys.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 3]
```diff
@@ -215,10 +225,23 @@ def analyze_gdn_decay(
     *,
     epsilon: float = 1e-3,
     static_gate_input: float = -0.3,
+    decay_parameter_storage_dtype: _DecayParameterStorageDtype | None = None,
 ) -> dict[str, list[float]]:
-    """Return deterministic per-head horizons for every GDN module in a model."""
+    """Return per-head horizons, optionally canonicalized to a checkpoint storage dtype."""
+    storage_dtype = None
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:239]: Fixed in #2394. The public boundary now type-checks before membership lookup, so unhashable values such as lists produce the documented ValueError instead of leaking TypeError. Added regressions for both unsupported strings and unhashable inputs.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 4]
```diff
@@ -215,10 +225,23 @@ def analyze_gdn_decay(
...
+    if decay_parameter_storage_dtype is not None:
+        try:
+            storage_dtype = _STORAGE_DTYPES[decay_parameter_storage_dtype]
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py]: Fixed in #2394. epsilon and static_gate_input are now validated at analyze_gdn_decay() before model discovery, with nonnumeric values normalized to the public ValueError contract. Added finite/range and unhashable-input regressions.

File: modelopt/torch/sparsity/state_sparsity/policy.py [hunk 5]
```diff
@@ -191,9 +202,8 @@ def _analyze_gdn_modules(
         a_log = module.A_log
         dt_bias = module.dt_bias
         try:
-            if not a_log.dtype.is_floating_point or not dt_bias.dtype.is_floating_point:
-                raise ValueError("GDN A_log and dt_bias must use floating-point dtypes")
             if storage_dtype is not None:
+                _validate_gdn_decay_tensors(a_log, dt_bias)
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/policy.py:213]: Fixed in #2394. _analyze_gdn_modules() validates the live decay tensors before any checkpoint-storage cast, so casting cannot hide an invalid integer source dtype; compute_gdn_decay_horizons still validates the storage-canonical tensors after casting.

File: modelopt/torch/sparsity/state_sparsity/conversion.py [hunk 6]
```diff
@@ -108,16 +108,23 @@ def restore_dasc_model(model: nn.Module, config: DASCConfig, metadata: MetadataD
...
         validate_dasc_model_structure(model, policy)
-        validate_dasc_decay_parameters(model, policy)
-    except ApplyModeError as error:
+    except _DASCModelStructureMismatchError as error:
         warnings.warn(
             f"{error}. The saved DASC policy is stale; re-run calibrate() before deployment",
             stacklevel=2,
         )
+    else:
+        try:
```
[the remark this answers is not in the document]
Comment (kaix-nv) [reply modelopt/torch/sparsity/state_sparsity/conversion.py:121]: Fixed in #2394. A supported GDN identity whose decay tensors are temporarily unavailable is now a typed recoverable-staleness condition on both save and restore. Removing/replacing the supported architecture remains a hard ApplyModeError. The lifecycle boundary is documented and covered by a save/restore/export regression.
